ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStoreV2.java

```diff
@@ -235,17 +237,67 @@ public void updateTtl(List<MessageInfo> messageInfos) {
     // TODO
   }
 
+  /**
+   * Finds keys absent in the Store from given list of keys
+   * @param storeKeys Input list of keys to check for existence
+   * @return Set of keys absent in the Store
+   */
   @Override
   public Set<StoreKey> findMissingKeys(List<StoreKey> storeKeys) {
-    // TODO
-    return new HashSet<>();
+    Set<StoreKey> missingKeys = new HashSet<>();
+    Timer.Context timer = azureMetrics.missingKeysQueryTime.time();
+    try {
+      for (StoreKey storeKey : storeKeys) {
+        try {
+          if (!blobContainerClient.getBlobClient(storeKey.getID()).exists()) {
+            missingKeys.add(storeKey);
+          } // if
+        } catch (Exception e) {
+          azureMetrics.findKeyErrorCount.inc();
+          logger.error("Failed to find key {} in Azure blob storage because {}", storeKey.getID(), e.getMessage());
+          throw e;
+        } // try
+      } // for
+    } finally {
+      timer.stop();
+    } // try
+    return missingKeys;
+  }
+
+  public static final String DELETE_TIME_MS = "deleteTimeMs";
+  public static final String TTL_UPDATE_TIME_MS = "ttlUpdateTimeMs";
+  public static final String UNDELETE_TIME_MS = "undeleteTimeMs";
+
+  /**
+   * @return a {@link HashMap} of metadata key-value pairs.
+   */
+  public MessageInfo messageInfoFromMap(StoreKey storeKey, Map<String, String> map) {
+    return new MessageInfo(storeKey, Long.parseLong(map.get(SIZE)), map.containsKey(DELETE_TIME_MS),
```

Comment:
If the blob is deleted and then undeleted, isDeleted should be true or false?
map.containsKey(DELETE_TIME_MS) will be true even it's undeleted, right?

Reply:
This depends on how I implement delete() and undelete(), which is next PR. find() will just report what it finds. 
I think that if a blob deleted and then undeleted, the delete_timestamp should be cleared or removed from metadata and the lifeVersion must be incremented. isDeleted will be false.

Reply:
Can we find out what do we do in the Ambry server? Do we set isDelete to true or false when the blob is deleted and undeleted? If you want to delay this, please add TODO.

Anything we are uncertain at this moment, I'll suggest to add one specific TODO. Otherwise, when these questions pile up, we'll lose track of them. 
For example, TODO_VCR_BACKUP_V2, something like that and keep using the same TODO for all your PRs.

Reply:
Ok, I have a few thoughts and questions about delete/undelete. I will setup a meeting or add to my doc and discuss.